Nissan Leaf Integration (Carwings / NissanConnect EV)

filcole opened the pull request:
## Description:

Adds support for monitoring and controlling the Nissan Leaf remote control platform.

Currently introduces two switches (climate control and charging, charging can only be turned on as per the API limitation), 3 sensors (battery charge and range with and without AC on), a binary sensor for if the car is currently plugged in, device tracker for the car location (if the car is capable).

I'm based in the UK with a 2016 30kWh leaf so assistance from others is very useful.

Example entry for configuration.yaml:

```yaml
nissan_leaf:
  username: "username"
  password: "password"
  region: 'NE'     # Indicates Europe, must be changed for other regions
  nissan_connect: true  # Set to false for early Leafs 24kWh that don't report location
  update_interval: 
    hours: 1
  update_interval_charging: 
    minutes: 15
  update_interval_climate: 
    minutes: 5
  force_miles: true
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [X] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [X] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [X] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

https://github.com/home-assistant/home-assistant.io/pull/8073

Commit:
Added work so far.

Commit:
Change interval so nobody drains their battery when I put this online

Commit:
Added the warning notice.

Commit:
Async setup

Commit:
Still broken, but we're getting there.

Commit:
Back to synchronous, moved refresh stuff into DataStore

Commit:
Functional sensors!

Commit:
Added working switches, tweaked intervals a bit

Commit:
Fixed turn off result

Commit:
Moved plug status to binary_sensor, added smart intervals

Commit:
Documentation and car nickname stuff

Commit:
Syntax fixes and coveragerc additions

Commit:
Style fixes

Commit:
Fixing the final line length

Commit:
Fixed an issue with newer models and bad climate data

Commit:
Forgot to check my line endings.

Commit:
New icons for most of the components

Commit:
Hotfix for handling Nissan's awful servers

filcole committed:
Merge in work completed by BenWoodford

filcole committed:
Merge branch 'BenWoodford-nissanleaf' into nissanleaf

filcole committed:
Merge in fixes made by Phil Cole

Remove invalid FIXMEs and update TODOs
Fixes for pylint and test for CarwingsError exception rather than Exception
Flake8 fixes
Add pycarwings2 to requirements_all.txt
Add extra configuration documentation.
Use pycarwings2 from pip. Check server dates between requests.
Add sensor device class for battery.
Async conversion fixes
flake8 fixes and docstrings
Non-async charging is OK
Handle multiple cars in the configuration
Convert to async. Better imports for platforms
Fix scanning interval & prevent extra refreshes.  async switchover
Check discovery_info to prevent load of platforms
Ensure update frequency is always above a minimum interval (1 min).
Platforms don't have return values
Use values() instead of items() when not using key
Use snake_case (LeafCore becomes leaf_core)

commit 418b6bbcc49cf2909aac85869440435410abf3fd

filcole committed:
Add pycarwings2 to requirements_all.txt

filcole committed:
Make stopping charge error an 'info'. Remove TODO.

filcole committed:
Request update from car after sending start charging command.

filcole committed:
Delay initial (slow) update for 15 seconds and make async

filcole committed:
Flake8 line length fixes

filcole committed:
Try to fix D401 'imperative mood' git diff tox errors

filcole committed:
Try to fix more D401 'imperative mood' tox errors

filcole committed:
Default interval of an hour in code, to match comments.

filcole committed:
Update to pycarwings2 2.3

filcole committed:
Update to pycarwings2 2.3 in requirements_all.txt

filcole committed:
Remove documentation, instead refering to home-assistant.io

BenWoodford commented:
In case there was any need to say it, @filcole does have my permission to submit the Leaf component as I don't really have much time to work on it these days.

dgomes commented on homeassistant/components/nissan_leaf.py:
```diff
@@ -0,0 +1,489 @@
+from homeassistant.helpers.event import async_track_point_in_utc_time
+from homeassistant.util.dt import utcnow
+
+# If testing then use the following kinds of URLs
```
These comments should not belong to the released version

dgomes commented on homeassistant/components/nissan_leaf.py:
```diff
@@ -0,0 +1,489 @@
+
+
+@asyncio.coroutine
+def async_setup(hass, config):
```
You need to review this function, it is way more complex then it needs to be.

look at the BMW_connect_drive component for an example

filcole replied:
@dgomes I'm unsure how to make this simpler.  Apart from this change request I think all other issues are resolved.

dgomes commented on homeassistant/components/nissan_leaf.py:
```diff
@@ -0,0 +1,489 @@
+
+        # The 12V battery is used when communicating with Nissan servers.
+        # The 12V battery is charged from the traction battery when not
+        # connected # and when the traction battery has enough charge. To
```
there is a # hanging loose

dgomes commented on homeassistant/components/nissan_leaf.py:
```diff
@@ -0,0 +1,489 @@
+
+        if battery_response is not None:
+            _LOGGER.debug("Battery Response: ")
+            _LOGGER.debug(battery_response.__dict__)
```
Don't split logs into 2 lines...

_LOGGER.debug("Battery Response: %s", battery_response.__dict__)

dgomes commented on homeassistant/components/nissan_leaf.py:
```diff
@@ -0,0 +1,489 @@
+            _LOGGER.debug(battery_response.__dict__)
+
+            if battery_response.answer['status'] == 200:
+                if int(battery_response.battery_capacity) > 100:
```
Don't understand this code  🤷‍♂️ , hope it is right 😅

dgomes commented on homeassistant/components/nissan_leaf.py:
```diff
@@ -0,0 +1,489 @@
+
+        if server_response is not None:
+            _LOGGER.debug("Server Response: ")
+            _LOGGER.debug(server_response.__dict__)
```
See above

dgomes commented on homeassistant/components/nissan_leaf.py:
```diff
@@ -0,0 +1,489 @@
+                    server_response.cruising_range_ac_off_km
+                )
+                self.signal_components()
+                self.last_battery_response = utcnow()
```
This code is tase same as the previous one, probably can be refactored ?

filcole replied:
Should be fixed by commit https://github.com/home-assistant/home-assistant/pull/19786/commits/33fcd3f60e2cda2ba7b25aeb8d0b39c892ea008a

dgomes replied:
I was referring to all the code that checks battery_response.

filcole replied:
Yes, sorry. Thanks. Since the nissan servers have changed the way they respond I've simplified the code and only consider them returning via one route, so we only have to handle it one type of response in `async_refresh_data` and `async_get_battery`.

dgomes commented on homeassistant/components/nissan_leaf.py:
```diff
@@ -0,0 +1,489 @@
+            climate_response = await self.async_get_climate()
+            if climate_response is not None:
+                _LOGGER.debug("Got climate data for Leaf.")
+                _LOGGER.debug(climate_response.__dict__)
```
See above (no double logging)

dgomes commented on homeassistant/components/nissan_leaf.py:
```diff
@@ -0,0 +1,489 @@
+                    self.data[DATA_LOCATION] = location_response
+
+                    _LOGGER.debug("Location Response: ")
+                    _LOGGER.debug(location_response.__dict__)
```
See above

dgomes replied:
still double logging... make it:

_LOGGER.debug("Location Response: %s", location_response.__dict__)

dgomes commented on homeassistant/components/nissan_leaf.py:
```diff
@@ -0,0 +1,489 @@
+        self.request_in_progress = False
+        self.signal_components()
+
+    def signal_components(self):
```
This is a one line... probably could live without it :)

filcole replied:
Thanks. Fixed in https://github.com/home-assistant/home-assistant/pull/19786/commits/33fcd3f60e2cda2ba7b25aeb8d0b39c892ea008a

dgomes commented on homeassistant/components/nissan_leaf.py:
```diff
@@ -0,0 +1,489 @@
+        )
+
+        # Store the date from the nissan servers
+        start_date = start_server_info.answer[
```
I would probably add a try ... catch around this (just incase the server responds something different in the future)

filcole replied:
I've put a try/catch around the whole of async_get_battery.

dgomes commented on homeassistant/components/nissan_leaf.py:
```diff
@@ -0,0 +1,489 @@
+
+        if climate_result is not None:
+            _LOGGER.debug("Climate result:")
+            _LOGGER.debug(climate_result.__dict__)
```
See above

dgomes commented on homeassistant/components/sensor/nissan_leaf.py:
```diff
@@ -0,0 +1,121 @@
+        """Battery range in miles or kms."""
+        ret = 0
+
+        if self.ac_on is True:
```
no need for `is True`

dgomes replied:
it's still here :)

dgomes commented on homeassistant/components/sensor/nissan_leaf.py:
```diff
@@ -0,0 +1,121 @@
+        else:
+            ret = self.car.data[DATA_RANGE_AC_OFF]
+
+        if (self.car.hass.config.units.is_metric is False or
```
see above

dgomes commented on homeassistant/components/sensor/nissan_leaf.py:
```diff
@@ -0,0 +1,121 @@
+    @property
+    def unit_of_measurement(self):
+        """Battery range unit."""
+        if (self.car.hass.config.units.is_metric is False or
```
see above

dgomes commented on homeassistant/components/sensor/nissan_leaf.py:
```diff
@@ -0,0 +1,121 @@
+        if (self.car.hass.config.units.is_metric is False or
+                self.car.force_miles is True):
+            return "mi"
+        return "km"
```
use LENGTH_KILOMETERS and LENGTH_MILES

dgomes commented on homeassistant/components/switch/nissan_leaf.py:
```diff
@@ -0,0 +1,108 @@
+        devices.append(LeafClimateSwitch(value))
+
+    _LOGGER.debug("Calling add_devices for switches")
+    add_devices(devices, True)
```
We previously discussed logging... this is too much

dgomes commented on homeassistant/components/switch/nissan_leaf.py:
```diff
@@ -0,0 +1,108 @@
+    @property
+    def is_on(self):
+        """Return true if climate control is on."""
+        return self.car.data[DATA_CLIMATE] is True
```
remove `is True`

dgomes replied:
still there...

dgomes reviewed (changes requested):
I advise you to follow the pattern of other car platforms (bmw, volvo, chevy)

filcole commented on homeassistant/components/binary_sensor/nissan_leaf.py:
```diff
@@ -0,0 +1,55 @@
+        """Sensor name."""
+        return self.car.leaf.nickname + " Plug Status"
+
+    def log_registration(self):
```
`log_registration()` is called by `async_added_to_hass` in the `LeafEntity` class.  It's only purpose display which type of inherited entity, so will try to rework to use `self.__class__.__name__`

filcole committed:
Remove unneeded dispatcher_send()

filcole commented on homeassistant/components/device_tracker/nissan_leaf.py:
```diff
@@ -0,0 +1,46 @@
+                icon='mdi:car')
+
+    dispatcher_connect(hass, SIGNAL_UPDATE_LEAF, see_vehicle)
+    dispatcher_send(hass, SIGNAL_UPDATE_LEAF)
```
Thanks. Removed `dispatcher_send` line.

filcole replied:
Removed in commit https://github.com/home-assistant/home-assistant/pull/19786/commits/33fcd3f60e2cda2ba7b25aeb8d0b39c892ea008a and  https://github.com/home-assistant/home-assistant/pull/19786/commits/e4f61650f609e83c08f8d52510d7d36b4cd33e72

filcole committed:
Remove unneeded requirements comments

filcole committed:
Combine excess debugging.

filcole committed:
Remove single line method signal_components()

filcole commented:
@dgomes thank you for the review. The NissanConnectEV web service is currently down when it returns I'll checkin some changes to fix issues you identified.

filcole commented:
Nissan Servers are still not working for me. A migration Nissan has made between servers has corrupted/lost credentials for a number of users, see [this thread](https://www.speakev.com/threads/nissan-connect-update.132118).  Will resume on this when I can regain access to my car remotely as currently unable to test!

dgomes commented:
Don't worry, there is a huge backlog of PR's and one more review of your PR is needed before it can be merged. So take your time.

filcole committed:
Bump to version 2.4 of pycarwings2

filcole committed:
Remove unused dispatcher_send

filcole committed:
Simplify logging of LeafEntity registration

filcole committed:
Update requirements_all.txt

filcole committed:
Multiple changes

Increase timeout to 30 seconds
Only consider battery_status
Fix plugged in status
Better attempts at try/exception handling

filcole committed:
Fix line length

filcole committed:
Use pycarwings 2.5

filcole committed:
Remove pointless 'is True'

filcole committed:
Remove unnecessary 'is True/False'

filcole committed:
Remove unnecessary 'is True/False'

filcole committed:
Use LENGTH_MILES and LENGTH_KILOMETERS

filcole committed:
Remove excess logging in setup_platform()

filcole committed:
Remove unnecessary 'is True'

filcole committed:
Use pycarwings2 version 2.6

BenWoodford commented on homeassistant/components/nissan_leaf.py:
```diff
@@ -0,0 +1,467 @@
+            _LOGGER.debug("Server Response: %s", server_response.__dict__)
+
+            if server_response.answer['status'] == 200:
+                self.data[DATA_BATTERY] = server_response.battery_percent
```
Are you testing this with an actual Nissan Connect account with linked Leaf? What's the Server Response output when you enable debug logging?

stbenjam replied:
Yup, I do.  This is a 2018 LEAF SV.  This is the server response, battery_percent is a string: 

https://gist.githubusercontent.com/stbenjam/63cf03bd862d76f82f5ce34ca61a4a14/raw/52d20fc2ea90e5998d63923683df6e30baa9101a/server_response.log

BenWoodford replied:
How very Nissan...

filcole committed:
Require pycarwings2 version 2.7.

filcole committed:
Increase sleep delay for climate and location reponses.

dgomes commented on homeassistant/components/switch/nissan_leaf.py:
```diff
@@ -0,0 +1,108 @@
+        """Charging switch icon."""
+        if self.car.data[DATA_CHARGING]:
+            return 'mdi:flash'
+        return 'mdi:flash-off'
```
I'm not sure this is allowed (dynamic change of icons)

filcole replied:
There's a couple of examples of changing the Icon here:  https://github.com/home-assistant/home-assistant/blob/91aa874c0cc8b4e0daf5a268bc2d3ac4c2339fc2/homeassistant/components/sensor/dsmr.py#L254  and https://github.com/home-assistant/home-assistant/blob/cb07ea0d60acec43778687282504fdf8eb712885/homeassistant/components/switch/android_ip_webcam.py#L90 

Perhaps the `mdi:flash` and `mdi:flash` strings should be constants.

dgomes commented on homeassistant/components/nissan_leaf.py:
```diff
@@ -0,0 +1,467 @@
+
+        _LOGGER.info("Successfully logged in and fetched Leaf info")
+        _LOGGER.info(
+            "WARNING: This may poll your Leaf too often, and drain the 12V"
```
it's says WARNING.... maybe should be _LOGGER.warning

filcole replied:
This changed in commit https://github.com/home-assistant/home-assistant/pull/19786/commits/72924b812409f026ae6a7ec694f44c0cd9e1c14a

dgomes commented on homeassistant/components/nissan_leaf.py:
```diff
@@ -0,0 +1,467 @@
+    def device_state_attributes(self):
+        """Return default attributes for Nissan Leaf entities."""
+        return {
+            'homebridge_serial': self.car.leaf.vin,
```
what are these attributes for ?

filcole replied:
@BenWoodford Are you able to help with this one.  I've not got a homebridge, so I'm not sure.

dgomes replied:
we'll if this is for integration with home kit, this simply does not belong in the nissan leaf component.

filcole replied:
Removed homebridge_* state attributes in commit https://github.com/home-assistant/home-assistant/pull/19786/commits/cf0b761be652865a4bc96c53000addd29d53b0a2

dgomes commented on homeassistant/components/sensor/nissan_leaf.py:
```diff
@@ -0,0 +1,116 @@
+    @property
+    def device_class(self):
+        """Return the device class of the sensor."""
+        return 'battery'
```
use DEVICE_CLASS_BATTERY from const

filcole replied:
Thanks. Fixed in commit https://github.com/home-assistant/home-assistant/pull/19786/commits/72924b812409f026ae6a7ec694f44c0cd9e1c14a

dgomes commented on homeassistant/components/sensor/nissan_leaf.py:
```diff
@@ -0,0 +1,116 @@
+    @property
+    def name(self):
+        """Update sensor name depending on AC."""
+        if self.ac_on is True:
```
Name of an entity can't be dynamic! It breaks any automation

filcole replied:
`ac_on` is only set during the constructor of a LeafRangeSensor.  There's two sensors, one for the range with air conditioning off, and one for air conditioning on. Both or which are always instantiated. I will change `self.ac_on` to `self._ac_on` to indicate it's an internal property.

filcole replied:
Change made to `_ac_on` in commit https://github.com/home-assistant/home-assistant/pull/19786/commits/3c7c509506dfa4e39afed8eecfbd468569fba004

dgomes commented on homeassistant/components/nissan_leaf.py:
```diff
@@ -0,0 +1,467 @@
+
+        for component in LEAF_COMPONENTS:
+            if (component != 'device_tracker' or
+                    car_config[CONF_NCONNECT] is True):
```
remove "is True"

filcole replied:
Thanks. Fixed in commit https://github.com/home-assistant/home-assistant/pull/19786/commits/50034e67a4d8ed324b638cb0bcf94de3f2da7829

dgomes commented on homeassistant/components/nissan_leaf.py:
```diff
@@ -0,0 +1,467 @@
+    hass.data[DATA_LEAF] = {}
+    tasks = [async_setup_leaf(car) for car in config[DOMAIN]]
+    if tasks:
+        yield from asyncio.wait(tasks, loop=hass.loop)
```
Correct syntax (Python 3.5+):
```suggestion
        await asyncio.wait(tasks, loop=hass.loop)
```

filcole replied:
Fixed with commit https://github.com/home-assistant/home-assistant/pull/19786/commits/60b91781f18dc7c9c889a49902af8fb60aaba196

dgomes commented on homeassistant/components/sensor/nissan_leaf.py:
```diff
@@ -0,0 +1,116 @@
+    @property
+    def icon(self):
+        """Nice icon for range."""
+        return 'mdi:speedometer'
```
move to top of the file and use a constant

filcole replied:
Thanks. Fixed by commit https://github.com/home-assistant/home-assistant/pull/19786/commits/c09752f9a6e57321c0201f1227c0fe224d2d84cf

dgomes commented:
You should have a look at https://github.com/home-assistant/home-assistant/pull/19948 and make the proper changes.

filcole committed:
Remove unnecessary 'is True'

filcole commented on homeassistant/components/switch/nissan_leaf.py:
```diff
@@ -0,0 +1,108 @@
+    @property
+    def is_on(self):
+        """Return true if charging."""
+        return self.car.data[DATA_CHARGING] is True
```
Thanks. Fixed in commit https://github.com/home-assistant/home-assistant/pull/19786/commits/5dee035568ce098fd578ef4e6c07e7dc5e946540

filcole committed:
Increase frequent polling warning to _LOGGER.warning()

filcole committed:
Use DEVICE_CLASS_BATTERY

filcole committed:
Remove extraneous 'is True'.

filcole committed:
Move icon strings to constants.

filcole committed:
Remove unneeded key.

filcole committed:
LeafRangeSensor ac_on property is internal.

filcole committed:
Flake8 missing line

filcole commented on homeassistant/components/sensor/nissan_leaf.py:
```diff
@@ -0,0 +1,118 @@
+
+ICON_RANGE = 'mdi:speedometer'
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
```
Fixed by https://github.com/home-assistant/home-assistant/pull/19786/commits/9ce043011371622adaca21cdee65a6c9a37063c3

filcole committed:
Remove homebridge attributes.

filcole committed:
Merge branch 'dev' into nissanleaf

filcole committed:
Remove round battery % and range to whole numbers

filcole committed:
Use pycarwings2 2.8

filcole committed:
Move to embedded component model

filcole committed:
Reduce maximum attempts to 10 (5 mins)

filcole committed:
Include attempt count in 'waiting' log message

filcole commented:
Hi @dgomes,

> You should have a look at #19948 and make the proper changes.

I think I've made the changes necessary to meet #19948 in commit https://github.com/filcole/home-assistant/commit/676c9ef2c483d4ad3d18c42a6e6220d7b2ff6325

filcole committed:
Use await instead of yield. Remove @asyncio.coroutine decorators.

filcole committed:
Add @filcole as nissan_leaf codeowner

filcole committed:
Sync up to latest dev

filcole committed:
Fix checking for if not data returned from vehicle. Don't double send signal on location update.

filcole committed:
Exposed updated_on, update_in_progress and next_update attributes.

filcole committed:
Add nissan_leaf.update service that triggers an update.

filcole committed:
Flake8 line fixes

dgomes commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,509 @@
+            return False
+
+        _LOGGER.info("Successfully logged in and fetched Leaf info")
+        _LOGGER.warning(
```
Don't double log... even if at different levels (info - warning)

filcole replied:
Thanks.  I removed the _LOGGER.info("Successfully....") logging in commit https://github.com/home-assistant/home-assistant/pull/19786/commits/c4ced4b14b56fe0fe65d3e5caedb3299dc15252d

dgomes reviewed (approved):
I'll leave to other reviewers to have a second go :)

filcole committed:
Remove excess and double logging.

filcole committed:
Add updated_on attribute for device tracker.

filcole committed:
Fix crash if pycarwings2 doesn't provide cruising ranges.

fabaff committed:
Minor changes

fabaff committed:
Minor changes

fabaff committed:
Minor changes

fabaff committed:
Minor changes

fabaff committed:
Minor changes

klaasnicolaas commented:
Release 0.88 or 0.89?

stbenjam commented:
Thank you so much @filcole !!!!

bachya commented:
@klaasnicolaas This has been marked for 0.88.

MartinHjelmare commented:
No, this will go out in 0.89.

MartinHjelmare commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+            # This might need to be made async (somehow) causes
+            # homeassistant to be slow to start
+            await hass.async_add_job(leaf_login)
+        except(RuntimeError, urllib.error.HTTPError):
```
`RuntimeError` is a broad exception we should avoid to catch. Why do we need to do that?

MartinHjelmare commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+                _LOGGER.info("No start date from servers. Aborting")
+                return None
+
+            _LOGGER.info("Start server date=%s", start_date)
```
Debug max.

filcole replied:
Fixed with commit https://github.com/filcole/home-assistant/commit/119b2a3b20db6ea4cf05fda29921930c5e27bc86

MartinHjelmare commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+            _LOGGER.info("Start server date=%s", start_date)
+
+            # Request battery update from the car
+            _LOGGER.info("Requesting battery update, %s", self.leaf.vin)
```
Debug.

filcole replied:
Fixed with commit https://github.com/filcole/home-assistant/commit/119b2a3b20db6ea4cf05fda29921930c5e27bc86

MartinHjelmare commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+        from pycarwings2 import CarwingsError
+        try:
+            # First, check nissan servers for the latest data
+            start_server_info = await self.hass.async_add_job(
```
Replace with `hass.async_add_executor_job`.

MartinHjelmare commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+
+            # Request battery update from the car
+            _LOGGER.info("Requesting battery update, %s", self.leaf.vin)
+            request = await self.hass.async_add_job(self.leaf.request_update)
```
See above.

MartinHjelmare commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+                return None
+
+            for attempt in range(MAX_RESPONSE_ATTEMPTS):
+                _LOGGER.info("Waiting %s seconds for battery update (%s) (%s)",
```
Debug.

filcole replied:
Fixed with commit https://github.com/filcole/home-assistant/commit/119b2a3b20db6ea4cf05fda29921930c5e27bc86

MartinHjelmare commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+
+                # Note leaf.get_status_from_update is always returning 0, so
+                # don't try to use it anymore.
+                server_info = await self.hass.async_add_job(
```
See above.

MartinHjelmare commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+                )
+
+                latest_date = self._extract_start_date(server_info)
+                _LOGGER.info("Latest server date=%s", latest_date)
```
Debug.

filcole replied:
Fixed with commit https://github.com/filcole/home-assistant/commit/119b2a3b20db6ea4cf05fda29921930c5e27bc86

MartinHjelmare commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+                if latest_date is not None and latest_date != start_date:
+                    return server_info
+
+            _LOGGER.info("%s attempts exceeded return latest data from server",
```
Debug.

filcole replied:
Fixed with commit https://github.com/filcole/home-assistant/commit/119b2a3b20db6ea4cf05fda29921930c5e27bc86

MartinHjelmare commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+        """Request climate data from Nissan servers."""
+        from pycarwings2 import CarwingsError
+        try:
+            request = await self.hass.async_add_job(
```
See above.

MartinHjelmare commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+        """Set climate control mode via Nissan servers."""
+        climate_result = None
+        if toggle:
+            _LOGGER.info("Requesting climate turn on for %s", self.leaf.vin)
```
Debug.

filcole replied:
Fixed with commit https://github.com/filcole/home-assistant/commit/119b2a3b20db6ea4cf05fda29921930c5e27bc86

MartinHjelmare commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+        climate_result = None
+        if toggle:
+            _LOGGER.info("Requesting climate turn on for %s", self.leaf.vin)
+            request = await self.hass.async_add_job(
```
See above.

MartinHjelmare commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+            )
+            for attempt in range(MAX_RESPONSE_ATTEMPTS):
+                if attempt > 0:
+                    _LOGGER.info("Climate data not in yet (%s) (%s). "
```
Debug.

filcole replied:
Fixed with commit https://github.com/filcole/home-assistant/commit/119b2a3b20db6ea4cf05fda29921930c5e27bc86

MartinHjelmare commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+                                 attempt, PYCARWINGS2_SLEEP)
+                    await asyncio.sleep(PYCARWINGS2_SLEEP)
+
+                climate_result = await self.hass.async_add_job(
```
See above.

MartinHjelmare commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+                    break
+
+        else:
+            _LOGGER.info("Requesting climate turn off for %s", self.leaf.vin)
```
Debug.

filcole replied:
Fixed with commit https://github.com/filcole/home-assistant/commit/119b2a3b20db6ea4cf05fda29921930c5e27bc86

MartinHjelmare commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+
+        else:
+            _LOGGER.info("Requesting climate turn off for %s", self.leaf.vin)
+            request = await self.hass.async_add_job(
```
See above.

filcole replied:
Fixed with commit https://github.com/filcole/home-assistant/commit/119b2a3b20db6ea4cf05fda29921930c5e27bc86

MartinHjelmare commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+            for attempt in range(MAX_RESPONSE_ATTEMPTS):
+                if attempt > 0:
+                    _LOGGER.info("Climate data not in yet (%s) (%s). "
+                                 "Waiting (%s) seconds.", self.leaf.vin,
```
We don't end logging messages with period.

filcole replied:
Fixed in commit https://github.com/filcole/home-assistant/commit/62cf3ea463b36bdd7dd25e5865c80b7b9d9dade9

MartinHjelmare commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+                                  attempt, PYCARWINGS2_SLEEP)
+                    await asyncio.sleep(PYCARWINGS2_SLEEP)
+
+                climate_result = await self.hass.async_add_job(
```
See above.

MartinHjelmare commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+
+    async def async_get_location(self):
+        """Get location from Nissan servers."""
+        request = await self.hass.async_add_job(self.leaf.request_location)
```
See above.

MartinHjelmare commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+                              attempt, PYCARWINGS2_SLEEP)
+                await asyncio.sleep(PYCARWINGS2_SLEEP)
+
+            location_status = await self.hass.async_add_job(
```
See above.

MartinHjelmare commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+        # If that completes OK then trigger a fresh update to pull the
+        # charging status from the car after waiting a minute for the
+        # charging request to reach the car.
+        result = await self.hass.async_add_job(self.leaf.start_charging)
```
See above.

MartinHjelmare commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+        async_dispatcher_connect(
+            self.car.hass, SIGNAL_UPDATE_LEAF, self._update_callback)
+
+    def _update_callback(self):
```
Decorate this with `@callback` imported from core.py.

filcole replied:
Fixed by commit https://github.com/filcole/home-assistant/commit/66e976beec6a00329683a1966eeae659b0abf07d

MartinHjelmare commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+
+    def _update_callback(self):
+        """Update the state."""
+        self.schedule_update_ha_state(True)
```
Change to use `self.async_schedule_update_ha_state`.

MartinHjelmare commented on homeassistant/components/nissan_leaf/binary_sensor.py:
```diff
@@ -0,0 +1,44 @@
+DEPENDENCIES = ['nissan_leaf']
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
```
Rename `add_devices` to `add_entities`.

MartinHjelmare commented on homeassistant/components/nissan_leaf/binary_sensor.py:
```diff
@@ -0,0 +1,44 @@
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up of a Nissan Leaf binary sensor."""
+    _LOGGER.debug(
+        "binary_sensor setup_platform, discovery_info=%s", discovery_info)
```
Remove this.

filcole replied:
Fixed with commit https://github.com/filcole/home-assistant/commit/119b2a3b20db6ea4cf05fda29921930c5e27bc86

MartinHjelmare commented on homeassistant/components/nissan_leaf/binary_sensor.py:
```diff
@@ -0,0 +1,44 @@
+        "binary_sensor setup_platform, discovery_info=%s", discovery_info)
+
+    devices = []
+    for key, value in hass.data[DATA_LEAF].items():
```
Please use more descriptive variable names than `key` and `value`. What are these variables?

filcole replied:
Fixed with commit https://github.com/filcole/home-assistant/commit/648929cb56df72a0a34fbfd550465f934f4fb6dc

MartinHjelmare commented on homeassistant/components/nissan_leaf/binary_sensor.py:
```diff
@@ -0,0 +1,44 @@
+    add_devices(devices, True)
+
+
+class LeafPluggedInSensor(LeafEntity):
```
Inherit from binary_sensor base entity imported from the binary_sensor base component.

filcole replied:
Fixed with commit https://github.com/filcole/home-assistant/commit/86dadb4147846afc6411cc9b44bd221ef35cb780

MartinHjelmare commented on homeassistant/components/nissan_leaf/binary_sensor.py:
```diff
@@ -0,0 +1,44 @@
+        return "{} {}".format(self.car.leaf.nickname, "Plug Status")
+
+    @property
+    def state(self):
```
We should not overwrite the `state` property for this class. Use `is_on` instead.

filcole replied:
Fixed with commit https://github.com/filcole/home-assistant/commit/86dadb4147846afc6411cc9b44bd221ef35cb780

MartinHjelmare commented on homeassistant/components/nissan_leaf/device_tracker.py:
```diff
@@ -0,0 +1,46 @@
+
+def setup_scanner(hass, config, see, discovery_info=None):
+    """Set up the Nissan Leaf tracker."""
+    _LOGGER.debug("Setting up Scanner (device_tracker) for Nissan Leaf, "
```
Remove this.

filcole replied:
Fixed with commit https://github.com/filcole/home-assistant/commit/119b2a3b20db6ea4cf05fda29921930c5e27bc86

MartinHjelmare commented on homeassistant/components/nissan_leaf/device_tracker.py:
```diff
@@ -0,0 +1,46 @@
+            dev_id = 'nissan_leaf_{}'.format(slugify(host_name))
+            if not value.data[DATA_LOCATION]:
+                _LOGGER.debug("No position found for vehicle %s", key)
+                return False
```
Nothing is checking this return value.

filcole replied:
Fixed with commit https://github.com/filcole/home-assistant/commit/341b9475147ef483fe429d1550dc2520b34a7d6b

MartinHjelmare commented on homeassistant/components/nissan_leaf/sensor.py:
```diff
@@ -0,0 +1,113 @@
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Sensors setup."""
+    _LOGGER.debug("setup_platform nissan_leaf sensors, discovery_info=%s",
```
Remove this.

filcole replied:
Fixed with commit https://github.com/filcole/home-assistant/commit/119b2a3b20db6ea4cf05fda29921930c5e27bc86

filcole replied:
Fixed with commit https://github.com/filcole/home-assistant/commit/119b2a3b20db6ea4cf05fda29921930c5e27bc86

MartinHjelmare commented on homeassistant/components/nissan_leaf/switch.py:
```diff
@@ -0,0 +1,99 @@
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Nissan Leaf switch platform setup."""
+    _LOGGER.debug(
```
Remove this.

filcole replied:
Fixed with commit https://github.com/filcole/home-assistant/commit/119b2a3b20db6ea4cf05fda29921930c5e27bc86

MartinHjelmare commented on homeassistant/components/nissan_leaf/switch.py:
```diff
@@ -0,0 +1,99 @@
+    @property
+    def device_state_attributes(self):
+        """Return climate control attributes."""
+        attrs = super(LeafClimateSwitch, self).device_state_attributes
```
`super().device_state_attributes`

filcole replied:
Fixed with commit https://github.com/filcole/home-assistant/commit/15ec267870b77943cec51ea87795a18e94a074b3

MartinHjelmare commented on homeassistant/components/nissan_leaf/switch.py:
```diff
@@ -0,0 +1,99 @@
+            self.car.data[DATA_CLIMATE] = False
+
+    @property
+    def icon(self):
```
Remove this.

filcole replied:
I don't understand the change needed here.

filcole replied:
> Remove this.

I don't understand the change needed here

MartinHjelmare replied:
The switch entity already has frontend representation tailored to that entity. We usually don't overwrite this property for switches.

MartinHjelmare commented on homeassistant/components/nissan_leaf/switch.py:
```diff
@@ -0,0 +1,99 @@
+        if await self.car.async_start_charging():
+            self.car.data[DATA_CHARGING] = True
+
+    def turn_off(self, **kwargs):
```
Remove this if we don't support it.

MartinHjelmare replied:
Maybe this shouldn't be a switch at all? Maybe a scene is better?

filcole replied:
Yes, switch is sub-optimal.  Perhaps sensor and service - discuss in new pull request.

MartinHjelmare reviewed (commented):
Please open a new PR where we can address the comments.

Generally I think there is too much logging and the update logic seems complicated.

Consider use of async vs sync api. Don't do I/O in event loop.

I'm looking forward to the next PR.

filcole commented:
> Please open a new PR where we can address the comments.
> 
> Generally I think there is too much logging and the update logic seems complicated.
> 
> Consider use of async vs sync api. Don't do I/O in event loop.
> 
> I'm looking forward to the next PR.

@MartinHjelmare  Thank you for the review, and also to @fabaff  and @dgomes. I will submit another pull request when I've tried to address the issues raised.  I hope you don't mind if I comment against the conversation points raised as I'm attempting to resolve them.

filcole commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+            data_store = hass.data[DATA_LEAF][vin]
+            async_track_point_in_utc_time(
+                hass, data_store.async_update_data, utcnow())
+            return True
```
Fixed in https://github.com/filcole/home-assistant/commit/ce5cf8c63cca598197717309354c0e0bf4827fd3

filcole commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+            return True
+
+        _LOGGER.debug("Vin %s not recognised for update", vin)
+        return False
```
Fixed in commit https://github.com/filcole/home-assistant/commit/ce5cf8c63cca598197717309354c0e0bf4827fd3

filcole commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+class LeafDataStore:
+    """Nissan Leaf Data Store."""
+
+    def __init__(self, leaf, hass, car_config):
```
Fixed with commit https://github.com/filcole/home-assistant/commit/2403c30ee237c06ef09f82522ebc2c199dffa3ef

filcole commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+        if (self.last_battery_response is not None and
+                self.data[DATA_CHARGING] is False and
+                self.data[DATA_BATTERY] <= RESTRICTED_BATTERY):
+            _LOGGER.info("Low battery so restricting refresh frequency (%s)",
```
Fixed with commit https://github.com/filcole/home-assistant/commit/119b2a3b20db6ea4cf05fda29921930c5e27bc86

BenWoodford commented:
When you say too much logging, do you mean logging using info or in general?

Part of the reason for so much logging is that Nissan break their API on a semi-regular basis, and sometimes only for certain people (like how they only broke battery charge for older cars that report in bars). They have basically no sense of DevOps for API versioning/maintenance so the logging is needed at least on trace to debug issues that only occur for certain people.

MartinHjelmare commented:
Both. Generally we don't want to log on info level in components. If it's not a warning or error, we should use debug.

There some more places where the debug logs only seem to be used to verify our own logic. That's not how we should use logging. If we need to verify our own logic, we should use tests.

It's ok to log the result of an api but we should probably not log that we called the api, and definitely not log internal component variable changes. Sounds good?

MartinHjelmare commented:
For more discussion I would prefer to have that in the new PR. It's ok to open that before it's completely ready.

filcole commented on homeassistant/components/nissan_leaf/__init__.py:
```diff
@@ -0,0 +1,514 @@
+
+        for component in LEAF_COMPONENTS:
+            if component != 'device_tracker' or car_config[CONF_NCONNECT]:
+                load_platform(hass, component, DOMAIN, {}, car_config)
```
Thanks Martin.  I reworked the component to be sync, rather than async.  Will submit another pull request for review.
